### PR TITLE
feat: Add code generator for generic resource tagging

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

This adds a code generator that creates a method `AddTag` for each managed resource that has a field `Spec.ForProvider.Tag` to satisfy the generic interface `Tagged`.

It enables us to replace each specific `tagger` initializer that is currently copy-pasted for every controller with a central component reducing code redundancy.

The generated code abstracts from (most) of the more over a dozen tagging implementations that exist today. The following tag implementations are covered by this generator:

* `map[string]string`
* `map[string]*string`
* `[]Tag  {string, string}`
* `[]*Tag {string, string}`
* `[]Tag  {*string, *string}`
* `[]*Tag {*string, *string}`
* `[]Tag  (with string key and *string value)`
* `[]*Tag  (with string key and *string value)`

Note: It does _not_ fix the issue that tags are defined differently for each resource because that is an issue with the underlying ACK generator and AWS SDK (see #1244).

This is a preliminary for #1306.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Unit tests

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/generate.go [hunk 1]
```diff
@@ -16,21 +17,24 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// NOTE(negz): See the below link for details on what is happening here.
-// https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module
+// // NOTE(negz): See the below link for details on what is happening here.
```
Comment (haarchri) [inline apis/generate.go:20]: @MisterMX  looks like we have double `// `now ?

File: pkg/controller/common/doc.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
+Copyright 2021 The Crossplane Authors.
```
Comment (chlunde) [inline pkg/controller/common/doc.go:2]: nit: use 2022 for all new files?

File: pkg/controller/common/doc.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package common contains shared controller logic.
+package common
```
Comment (chlunde) [inline pkg/controller/common/doc.go:18]: Could we call this something more specific like tagger or tagutil?
https://github.com/golang/go/wiki/CodeReviewComments#package-names

**Comment by chlunde** [comment]
looks awesome! I think we could add IsUpToDate helpers the same way later?

File: apis/kms/v1alpha1/tagged.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// AddTag adds a tag to this Key. If it already exists, it will be overwritten.
+// It returns true if the tag has been added/changed. Otherwise false.
+func (mg *Key) AddTag(key string, value string) bool {
```
Comment (chlunde) [inline apis/kms/v1alpha1/tagged.go:21]: is this one hand writtend? If so, add a comment why?

**Comment by chlunde** [comment]
@MisterMX did you also try some of the resource types in a real cluster?

This is a huge PR, it looks good to me but there might be a "needle in the haystack" here. Is there anything in particular you think we should look at, anything that's not tested or that you feel less confident about?

**Comment by artem-nefedov** [comment]
Will this fix inconsistency in tagging input? (like how most resources expect tags as an array, but some select ones, such as eks Cluster and NodeGroup, expect them to be a map)

**Comment by haarchri** [comment]
@artem-nefedov no

**Pull request closed without merging**